oauth2/src/main/java/com/linecorp/armeria/server/auth/oauth2/OAuth2TokenIntrospectionAuthorizer.java

```diff
@@ -0,0 +1,211 @@
+ * in business logic.
+ */
+public class OAuth2TokenIntrospectionAuthorizer implements Authorizer<OAuth2Token> {
+
+    /**
+     * Returns a newly created {@link OAuth2TokenIntrospectionAuthorizerBuilder}.
+     *
+     * @param introspectionEndpoint A {@link WebClient} to facilitate the Token Introspection request. Must
+     *                              correspond to the Token Introspection endpoint of the OAuth 2.0 system.
+     * @param introspectionEndpointPath A URI path that corresponds to the Token Introspection endpoint of the
+     *                                  OAuth 2.0 system.
+     */
+    public static OAuth2TokenIntrospectionAuthorizerBuilder builder(WebClient introspectionEndpoint,
+                                                                    String introspectionEndpointPath) {
+        return new OAuth2TokenIntrospectionAuthorizerBuilder(introspectionEndpoint, introspectionEndpointPath);
+    }
+
+    static final AttributeKey<Integer> ERROR_CODE = AttributeKey.valueOf("x-oauth2-error");
+    static final AttributeKey<String> ERROR_TYPE = AttributeKey.valueOf("x-oauth2-error-type");
+    static final String UNSUPPORTED_TOKEN_TYPE = "unsupported_token_type";
+    static final String INVALID_TOKEN = "invalid_token";
+    static final String INSUFFICIENT_SCOPE = "insufficient_scope";
+
+    private final Cache<String, OAuth2TokenDescriptor> tokenCache;
+    private final Set<String> permittedScope;
+    @Nullable
+    private final String accessTokenType;
+    @Nullable
+    private final String realm;
+    private final TokenIntrospectionRequest tokenIntrospectionRequest;
+    private final AuthFailureHandler authFailureHandler;
+
+    OAuth2TokenIntrospectionAuthorizer(Cache<String, OAuth2TokenDescriptor> tokenCache,
+                                       @Nullable String accessTokenType, @Nullable String realm,
+                                       Set<String> permittedScope,
+                                       TokenIntrospectionRequest tokenIntrospectionRequest) {
+        this.tokenCache = requireNonNull(tokenCache, "tokenCache");
+        this.accessTokenType = accessTokenType;
+        this.realm = realm;
+        this.permittedScope = requireNonNull(permittedScope, "permittedScope");
+        this.tokenIntrospectionRequest =
+                requireNonNull(tokenIntrospectionRequest, "tokenIntrospectionRequest");
+        authFailureHandler =
+                new OAuth2AuthorizationFailureHandler(accessTokenType, realm, String.join(" ", permittedScope));
+    }
+
+    /**
+     * Creates a new {@link AuthService} that authorizes HTTP requests using OAuth 2.0 Token Introspection.
+     */
+    public AuthService asAuthService(HttpService delegate) {
+        return AuthService.builder()
+                          .addOAuth2(this)
+                          .onFailure(authFailureHandler())
+                          .build(delegate);
+    }
```

Comment:
This is interesting. I was rather expecting something like:

```java
AuthService.builder()
           .addOAuth2(OAuth2TokenIntrospectionAuthorizer.builder()....build())
           .newDecorator(); // or .build(..);
```

One missing piece here is that you cannot specify the failure handler. To solve this problem, what do you think about adding some default methods to `Authorizer`? e.g.

```java
public interface Authorizer<T> {
    ...
    @Nullable // null means 'use the default' or what's specified in the builder.
    default AuthSuccessHandler successHandler() { return null; }
    @Nullable // null means use the default or what's specified in the builder.
    default AuthFailureHandler failureHandler() { return null; }
}
```

.. so that you can specify the success/failure handler specific to OAuth2.

Reply:
Sorry, did not understand your proposal. What exactly am I missing?

Reply:
I meant:

A usual pattern for creating an `AuthService` with an `Authorizer` is to pass an `Authorizer` when building an `AuthService` with an `AuthServiceBuilder`. What's suggested here is the opposite; a user calls `Authorizer.asAuthService()`. Could we change this so it's consistent with others?

Reply:
As you pointed, the usual pattern does not allow configuring the failure processing bound to the authorization logic. That's why I had to do it this way.
Do you suggest modifying `com.linecorp.armeria.server.auth.Authorizer` and `com.linecorp.armeria.server.auth.AuthServiceBuilder` to support success and failure handlers provided by the `Authorizer`?

Reply:
> Do you suggest modifying `com.linecorp.armeria.server.auth.Authorizer` and `com.linecorp.armeria.server.auth.AuthServiceBuilder` to support success and failure handlers provided by the `Authorizer`?

Exactly! (Sorry about a very late reply. 🙏 Was busy joining a new company 😉)

Reply:
No problem. I will follow your proposal and push an update in the next few days.